### ISSUE TITLE
[Question] Why should we call stream_bucket_make_writeable inside a loop?

### ISSUE THREAD

**Issue by CViniciusSDias** [clue/php-stream-filter#27]
Every single implementation of `php_user_filter` I see calls the `stream_bucket_make_writeable` function inside a loop but in every single test I made it doesn't make any difference. The second call to `stream_bucket_make_writeable` always returns `null` in my tests.

In which case is this loop really needed?

**Comment by clue** [comment]
Hi @CViniciusSDias, that's a good question!

For reference, I assume that what you're asking for deals with an implementation and nothing that is exposed to consumers of this library:

https://github.com/clue/php-stream-filter/blob/5a58cc30a8bd6a4eb8f856adf61dd3e013f53f71/src/CallbackFilter.php#L60-L65

[PHP's documentation](https://www.php.net/manual/en/function.stream-filter-register.php#refsect1-function.stream-filter-register-examples) suggests the `stream_bucket_make_writable()` should be called in a loop to get all "buckets" from a "bucket brigade". This suggests it may return multiple "chunks" for a single write. As such, I would rephrase the question: Why should we *not* call it in a loop?

I agree that the underlying stream filter API is somewhat *weird* and that's exactly the motivation for creating this project (https://github.com/clue/php-stream-filter#why). By relying on this project's API, you shouldn't have to worry about these details and can simply rely on a single function signature as documented.

I believe this has been answered, so I'm closing this for now. Please come back with more details if you feel this needs clarification and we can reopen this :+1:

**Comment by CViniciusSDias** [comment]
Hey, @clue ! Thanks for answering.
I won't reopen it but my question was more like: Have you ever encoutered a case where using it without the loop led to different results? Cause in every sing test I made it does't make any difference.
About using you project, I already do! I was studying further the filters API so I stumbled into this "problem". I actually recommended this package in an online course I recorded for Alura (https://alura.com.br). It's still not available but I just wanted to let you know! :-)

**Comment by clue** [comment]
@CViniciusSDias Thanks for getting back at this :-)
Tbqh I don't recall whether I've been able to reproduce this, but I'd figure it doesn't make much of a difference in terms of the code base anyway, so I'd rather be safe than sorry. Also note that this project still supports legacy PHP 5.3 (and legacy HHVM to some degree), so I think sticking to what PHP's documentation suggests makes even more sense.

**Comment by KennedyTedesco** [comment]
According to my tests, `while ($bucket = stream_bucket_make_writeable($in)) { ` reproduce the same as:

https://github.com/KennedyTedesco/simple-stream-filter/blob/0.1/src/CallbackFilter.php#L11-L13

The `filter()` function is hit more than once and the `stream_bucket_make_writeable()` doesn't return more than one bucket per time.

And, yeah, the documentation is a bit poor.

**Comment by clue** [comment]
@KennedyTedesco What you're saying may have applied to your tests, but it's in conflict what PHP's documentation says (multiple times):
* https://www.php.net/manual/en/function.stream-filter-register.php#refsect1-function.stream-filter-register-examples
* https://www.php.net/manual/en/php-user-filter.filter.php#refsect1-php-user-filter.filter-parameters

> in is a resource pointing to a bucket brigade which contains **one or more** bucket objects containing data to be filtered. [emphasis mine]

Tbqh, I'm not sure I follow what you're trying to achieve in this ticket, so let's approach this from another direction: What would be gained by removing the 3-line while loop and replacing it with a 3-line if statement?

**Comment by KennedyTedesco** [comment]
@clue You are right. No only documentation says, but also any single test about this in the PHP repository.

I just came out with this question (same as @CViniciusSDias ) when I was studying all the possibilities. Just trying to keep an open mind here, because every single test and use case I ever had, the `while()` wasn't touch more than once.

But you made the right choice following what the documentation says. Other libraries use the same approach.

I'm not saying I'll cut off the while (probably not), as I said before, I'm just exploring things.

@clue Thank you so much for your work and patience to answer the issues, it's inspiring!
